contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,728 @@
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
+        address indexed _recipient,
+        address indexed _setToken,
+        address indexed _inputToken,
+        uint256 _amountIn,
+        uint256 _amountOut
+    );
+    event ExchangeRedeem(
+        address indexed _recipient,
+        address indexed _setToken,
+        address indexed _outputToken,
+        uint256 _amountIn,
+        uint256 _amountOut
+    );
+
+    /* ============ Constructor ============ */
+
+    constructor(
+        address _uniFactory,
+        IUniswapV2Router02 _uniRouter, 
+        address _sushiFactory, 
+        IUniswapV2Router02 _sushiRouter, 
+        IBasicIssuanceModule _basicIssuanceModule
+    )
+        public
+    {
+        uniFactory = _uniFactory;
+        uniRouter = _uniRouter;
+
+        sushiFactory = _sushiFactory;
+        sushiRouter = _sushiRouter;
+
+        WETH = uniRouter.WETH();
+        basicIssuanceModule = _basicIssuanceModule;
+        IERC20(WETH).approve(address(uniRouter), PreciseUnitMath.maxUint256());
+        IERC20(WETH).approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+    
+    /* ============ Public Functions ============ */
+    
+    /**
+     * Runs all the necessary approval functions required for a given ERC20 token.
+     * This function can be called when a new token is added to a set token during a 
+     * rebalance.
+     *
+     * @param _token    Address of the token which needs approval
+     */
+    function approveToken(IERC20 _token) public {
+        _token.approve(address(uniRouter), MAX_UINT96);
+        _token.approve(address(sushiRouter), MAX_UINT96);
+        _token.approve(address(basicIssuanceModule), MAX_UINT96);
+    }
+
+    /* ============ External Functions ============ */
+    
+    /**
+     * Runs all the necessary approval functions required for a list of ERC20 tokens.
+     *
+     * @param _tokens    Addresses of the tokens which need approval
+     */
+    function approveTokens(IERC20[] calldata _tokens) external {
+        for(uint256 i = 0; i < _tokens.length; i++)
+            approveToken(_tokens[i]);
+    }
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a set token. This function need to be called only once before the first time
+     * this smart contract is used on any particular set token.
+     *
+     * @param _setToken    Address of the set token being initialized
+     */
+    function approveSetToken(ISetToken _setToken) external {
+        ISetToken.Position[] memory positions = _setToken.getPositions();
+        for (uint256 i = 0; i < positions.length; i++)
+            approveToken(IERC20(positions[i].component));
+    }
+
+    /**
+     * Issues set tokens for an exact amount of input token. 
+     * Acquires set token components at the best price accross uniswap and sushiswap.
+     * Uses the acquired components to issue the set tokens.
```

Comment:
`Issues SetTokens for an exact amount of input ERC20 token. The ERC20 token must be approved by the sender to this contract`.